pkg/agent/client/endpoint_resolver.go

```diff
@@ -0,0 +1,218 @@
+	"k8s.io/apiserver/pkg/util/proxy"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+)
+
+type Listener interface {
+	Enqueue()
+}
+
+type EndpointResolver struct {
+	name        string
+	namespace   string
+	serviceName string
+	servicePort int32
+	// serviceLister is used to retrieve the antrea Service when selecting one of the Service's Endpoints.
+	serviceLister       corev1listers.ServiceLister
+	serviceListerSynced cache.InformerSynced
+	// endpointLister is used to select and Endpoint for the antrea Service. This Endpoint will
+	// be used to access the Service without depending on the availability of a proxy (whether
+	// kube-proxy or AntreaProxy).
+	endpointsLister       corev1listers.EndpointsLister
+	endpointsListerSynced cache.InformerSynced
+	queue                 workqueue.RateLimitingInterface
+	listeners             []Listener
+	endpointURL           atomic.Pointer[url.URL]
+}
+
+func NewEndpointResolver(
+	namespace, serviceName string,
+	servicePort int32,
+	serviceInformer coreinformers.ServiceInformer,
+	endpointsInformer coreinformers.EndpointsInformer,
+) *EndpointResolver {
+	key := namespace + "/" + serviceName
+	controllerName := fmt.Sprintf("ServiceEndpointResolver-%s", key)
+	resolver := &EndpointResolver{
+		name:                  controllerName,
+		namespace:             namespace,
+		serviceName:           serviceName,
+		servicePort:           servicePort,
+		serviceLister:         serviceInformer.Lister(),
+		serviceListerSynced:   serviceInformer.Informer().HasSynced,
+		endpointsLister:       endpointsInformer.Lister(),
+		endpointsListerSynced: endpointsInformer.Informer().HasSynced,
+		queue:                 workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(100*time.Millisecond, 5*time.Second), controllerName),
+	}
+	serviceInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: func(obj interface{}) bool {
+			if service, ok := obj.(*corev1.Service); ok {
+				return service.Namespace == namespace && service.Name == serviceName
+			}
+			if tombstone, ok := obj.(cache.DeletedFinalStateUnknown); ok {
+				if service, ok := tombstone.Obj.(*corev1.Service); ok {
+					return service.Namespace == namespace && service.Name == serviceName
+				}
+			}
+			return false
+		},
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				resolver.queue.Add(key)
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				// This should not happen: both objects should be Services in the
+				// update event handler.
+				oldSvc, ok := oldObj.(*corev1.Service)
+				if !ok {
+					return
+				}
+				newSvc, ok := newObj.(*corev1.Service)
+				if !ok {
+					return
+				}
+				// We do not care about potential Status updates.
+				if reflect.DeepEqual(newSvc.Spec, oldSvc.Spec) {
```

Comment:
maybe just check generation which is more efficient

Reply:
Is Generation available for all core resources, I never know?

Reply:
I thought setting/updating generation is automatic for all resources, getting that impression from Egress and AntreaNetworkPolicy CRs, but just confirmed I was wrong: Service doesn't have a generation, the generation calculation of core resources is added case-by-case.